RingBuffer for stats

jage opened the pull request:
### WHAT is this pull request doing?

Speed up internal stats collection, a ring buffer should be faster for our use case.

While not the most common operation, for a broker with a lot of objects (queues, exchanges etc) we want to minimize any CPU time spent on just keeping track of the stats.

Improvements to the `to_a` method that's used in the UI/API, i.e. any time we refresh the web page with new stats (proposed to be once per second in #1195) should help a bit, even though the full JSON serialization is probably much more expensive. Getting around 9% improvements in the simulated http_benchmark.cr

See benchmark/ directory for all micro benchmarks (should be cleaned up before we merge this). Also kept the old Deque based code until we're ready to say this is good.

```
RingBuffer (steady-state) 355.76M (  2.81ns) (± 5.70%)  0.0B/op        fastest
     Deque (steady-state) 177.83M (  5.62ns) (± 2.41%)  0.0B/op   2.00× slower

RingBuffer (fill from empty) 285.64k (  3.50µs) (± 0.76%)  12.1kB/op        fastest
     Deque (fill from empty) 202.75k (  4.93µs) (± 1.55%)  12.1kB/op   1.41× slower

RingBuffer (to_a) 872.70k (  1.15µs) (± 0.84%)  12.1kB/op        fastest
     Deque (to_a) 146.77k (  6.81µs) (± 0.94%)  12.1kB/op   5.95× slower
```

---

Why RingBuffer is faster:

Bitwise Modulo Operations

- RingBuffer uses `& @mask` for wrap-around (single CPU instruction)
- Deque uses conditional branches (`if index >= @capacity`)
- Bitwise `AND` avoids branch misprediction penalties

No Memory Shifting

- RingBuffer: Simply overwrites oldest value when full
- Deque: Calls shift to remove first element, causing memory movement

Simpler Push Logic

- RingBuffer: Direct write to calculated position
- Deque: Check size → shift if needed → then push (3 operations vs 1)

CPU Cache Optimization

- RingBuffer: Predictable circular access pattern
- Deque: Shift operations invalidate cache lines

No Dynamic Memory Management

- RingBuffer: Fixed allocation, never grows
- Deque: Resize checks and potential reallocation overhead

### HOW can this pull request be tested?

Specs and looking at the stats manually, but we might want to improve specs.

carlhoerberg reviewed (commented):
Nice! Any reason the old `Stat` is kept?

jage committed:
RingBuffer for stats

Replace per-stat Deque(Float64)/Deque(UInt32) logs with a fixed-capacity
RingBuffer(T). The buffer rounds capacity up to a power of two for
bitmask indexing, backs numeric T with GC.malloc_atomic (no pointer
scanning) and other types with a scanned allocation, and includes
Enumerable(T) to avoid intermediate to_a allocations.

Drops the per-update shift loop in update_rates since the ring buffer
overwrites oldest entries automatically.

jage commented:
This is pretty CPU efficient but not good enough memory management. We're wasting more memory with with the power of 2 setup, 128 slots -> 1024 bytes which Boehm slot allocates 1344 byte for. Got something better that will replace this.
